src/thread/setns.rs

```diff
@@ -0,0 +1,122 @@
+#![allow(unsafe_code)]
+
+use core::convert::TryFrom;
+
+use bitflags::bitflags;
+
+use crate::backend::c::c_int;
+use crate::backend::thread::syscalls;
+use crate::fd::BorrowedFd;
+use crate::io;
+
+/// New time namespace.
+const CLONE_NEWTIME: u32 = 0x00_00_00_80;
+/// New mount namespace group.
+const CLONE_NEWNS: u32 = 0x00_02_00_00;
+/// New cgroup namespace.
+const CLONE_NEWCGROUP: u32 = 0x02_00_00_00;
+/// New utsname namespace.
+const CLONE_NEWUTS: u32 = 0x04_00_00_00;
+/// New ipc namespace.
+const CLONE_NEWIPC: u32 = 0x08_00_00_00;
+/// New user namespace.
+const CLONE_NEWUSER: u32 = 0x10_00_00_00;
+/// New pid namespace.
+const CLONE_NEWPID: u32 = 0x20_00_00_00;
+/// New network namespace.
+const CLONE_NEWNET: u32 = 0x40_00_00_00;
+
+bitflags! {
+    /// Thread name space type.
+    pub struct ThreadNameSpaceType: u32 {
+        /// Time name space.
+        const TIME = CLONE_NEWTIME;
+        /// Mount name space.
+        const MOUNT = CLONE_NEWNS;
+        /// Control group (CGroup) name space.
+        const CONTROL_GROUP = CLONE_NEWCGROUP;
+        /// `Host name` and `NIS domain name` (UTS) name space.
+        const HOST_NAME_AND_NIS_DOMAIN_NAME = CLONE_NEWUTS;
+        /// Inter-process communication (IPC) name space.
+        const INTER_PROCESS_COMMUNICATION = CLONE_NEWIPC;
+        /// User name space.
+        const USER = CLONE_NEWUSER;
+        /// Process ID name space.
+        const PROCESS_ID = CLONE_NEWPID;
+        /// Network name space.
+        const NETWORK = CLONE_NEWNET;
+    }
+}
+
+/// Type of name space referred to by a link.
+#[derive(Copy, Clone, Debug, Eq, PartialEq)]
+#[repr(u32)]
+pub enum LinkNameSpaceType {
+    /// Time name space.
+    Time = CLONE_NEWTIME,
+    /// Mount name space.
+    Mount = CLONE_NEWNS,
+    /// Control group (CGroup) name space.
+    ControlGroup = CLONE_NEWCGROUP,
+    /// `Host name` and `NIS domain name` (UTS) name space.
+    HostNameAndNISDomainName = CLONE_NEWUTS,
+    /// Inter-process communication (IPC) name space.
+    InterProcessCommunication = CLONE_NEWIPC,
+    /// User name space.
+    User = CLONE_NEWUSER,
+    /// Process ID name space.
+    ProcessID = CLONE_NEWPID,
+    /// Network name space.
+    Network = CLONE_NEWNET,
+}
+
+impl TryFrom<u32> for LinkNameSpaceType {
+    type Error = io::Errno;
+
+    fn try_from(value: u32) -> Result<Self, Self::Error> {
+        match value {
+            CLONE_NEWTIME => Ok(Self::Time),
+            CLONE_NEWNS => Ok(Self::Mount),
+            CLONE_NEWCGROUP => Ok(Self::ControlGroup),
+            CLONE_NEWUTS => Ok(Self::HostNameAndNISDomainName),
+            CLONE_NEWIPC => Ok(Self::InterProcessCommunication),
+            CLONE_NEWUSER => Ok(Self::User),
+            CLONE_NEWPID => Ok(Self::ProcessID),
+            CLONE_NEWNET => Ok(Self::Network),
+            _ => Err(io::Errno::RANGE),
+        }
+    }
+}
```

Comment:
rustix doesn't currently provide these kinds of `TryFrom` impls. Typically the way this works is that regular Rust users should always be using the enum type, so it don't need these, and mustang implements this conversion manually. Is there a need to do something different here?